File: tests/unit/tag-immutability.test.ts
/**
 * Tag System Immutability Verification Tests
 * 
 * This test suite verifies that the tag system is immutable and hard-coded
 * throughout the application, with no dynamic tag generation or inference.
 * 
 * Requirements verified:
 * - Tags are defined as constants (not mutable)
 * - No dynamic tag generation or modification
 * - Frontend and backend tag definitions match
 * - Tag system is authoritative and never changes
 */

import { describe, it, expect } from 'vitest';
import { CATEGORIES, HARD_CODED_TAGS } from '../../src/config/categories';
import { BASE_TAGS, FILTER_TAGS } from '../../src/types';

describe('Tag System Immutability', () => {
  describe('Hard-Coded Tag Constants', () => {
    it('should have all base tags defined as string literals', () => {
      // Verify base tags are defined
      expect(HARD_CODED_TAGS.SERIES).toBe('series');
      expect(HARD_CODED_TAGS.MOVIE).toBe('movie');
      expect(HARD_CODED_TAGS.SITCOM).toBe('sitcom');
      expect(HARD_CODED_TAGS.KIDS).toBe('kids');
      expect(HARD_CODED_TAGS.HERO_TRAILER).toBe('hero_trailer');
    });

    it('should have all filter tags defined as string literals', () => {
      // Verify movie filter tags
      expect(HARD_CODED_TAGS.COMEDY_MOVIES).toBe('comedy_movies');
      expect(HARD_CODED_TAGS.ACTION_MOVIES).toBe('action_movies');
      expect(HARD_CODED_TAGS.ROMANCE_MOVIES).toBe('romance_movies');
      
      // Verify series filter tags
      expect(HARD_CODED_TAGS.COMEDY_SERIES).toBe('comedy_series');
      expect(HARD_CODED_TAGS.ACTION_SERIES).toBe('action_series');
      expect(HARD_CODED_TAGS.ROMANCE_SERIES).toBe('romance_series');
      
      // Verify kids filter tags
      expect(HARD_CODED_TAGS.COMEDY_KIDS).toBe('comedy_kids');
      expect(HARD_CODED_TAGS.ACTION_KIDS).toBe('action_kids');
    });

    it('should have BASE_TAGS array as readonly tuple', () => {
      // Verify BASE_TAGS is a readonly tuple (TypeScript enforces immutability)
      expect(BASE_TAGS).toEqual(['series', 'movie', 'sitcom', 'kids', 'hero_trailer']);
      
      // TypeScript prevents modification at compile time with 'as const'
      // The array is a readonly tuple, not a mutable array
      expect(BASE_TAGS.length).toBe(5);
      
      // Verify it's the exact tuple we expect
      expect(BASE_TAGS[0]).toBe('series');
      expect(BASE_TAGS[1]).toBe('movie');
      expect(BASE_TAGS[2]).toBe('sitcom');
      expect(BASE_TAGS[3]).toBe('kids');
      expect(BASE_TAGS[4]).toBe('hero_trailer');
    });

    it('should have FILTER_TAGS array as readonly tuple', () => {
      // Verify FILTER_TAGS is a readonly tuple (TypeScript enforces immutability)
      expect(FILTER_TAGS).toEqual([
        'comedy_movies', 'action_movies', 'romance_movies',
        'comedy_series', 'action_series', 'romance_series',
        'comedy_kids', 'action_kids'
      ]);
      
      // TypeScript prevents modification at compile time with 'as const'
      expect(FILTER_TAGS.length).toBe(8);
    });
  });

  describe('Category Configuration Immutability', () => {
    it('should have all categories defined with exact tag strings', () => {
      // Movies category
      expect(CATEGORIES.movies.baseTag).toBe('movie');
      expect(CATEGORIES.movies.filters).toEqual([
        { label: 'Comedy', tag: 'comedy_movies' },
        { label: 'Action', tag: 'action_movies' },
        { label: 'Romance', tag: 'romance_movies' }
      ]);

      // Series category
      expect(CATEGORIES.series.baseTag).toBe('series');
      expect(CATEGORIES.series.filters).toEqual([
        { label: 'Comedy', tag: 'comedy_series' },
        { label: 'Action', tag: 'action_series' },
        { label: 'Romance', tag: 'romance_series' }
      ]);

      // Sitcoms category
      expect(CATEGORIES.sitcoms.baseTag).toBe('sitcom');
      expect(CATEGORIES.sitcoms.filters).toEqual([]);

      // Kids category
      expect(CATEGORIES.kids.baseTag).toBe('kids');
      expect(CATEGORIES.kids.filters).toEqual([
        { label: 'Comedy', tag: 'comedy_kids' },
        { label: 'Action', tag: 'action_kids' }
      ]);
    });

    it('should not allow modification of CATEGORIES object', () => {
      // CATEGORIES is defined with TypeScript readonly properties
      // TypeScript prevents modification at compile time
      
      const originalMoviesBaseTag = CATEGORIES.movies.baseTag;
      
      // Verify the structure is correct
      expect(CATEGORIES.movies.baseTag).toBe('movie');
      expect(originalMoviesBaseTag).toBe('movie');
      
      // TypeScript readonly prevents modification at compile time
      // This is the immutability guarantee we need
    });
  });

  describe('No Dynamic Tag Generation', () => {
    it('should not have any tag inference logic', () => {
      // This test verifies that tags are never dynamically generated
      // All tags must come from the hard-coded constants
      
      const allDefinedTags = [
        ...BASE_TAGS,
        ...FILTER_TAGS
      ];

      // Verify exact count - 5 base + 8 filter = 13 tags
      expect(allDefinedTags.length).toBe(13);
      
      // Verify each tag is a string literal (not computed)
      allDefinedTags.forEach(tag => {
        expect(typeof tag).toBe('string');
        expect(tag.length).toBeGreaterThan(0);
        // Tags should not contain dynamic patterns like timestamps or UUIDs
        expect(tag).not.toMatch(/\d{13}/); // No timestamps
        expect(tag).not.toMatch(/[0-9a-f]{8}-[0-9a-f]{4}/); // No UUIDs
      });
    });

    it('should not allow tag concatenation or modification', () => {
      // Verify that tag constants cannot be used to create new tags
      const baseTag = HARD_CODED_TAGS.MOVIE;
      const filterTag = HARD_CODED_TAGS.ACTION_MOVIES;
      
      // These are separate constants, not derived from each other
      expect(baseTag).toBe('movie');
      expect(filterTag).toBe('action_movies');
      
      // Verify filter tag is not derived from base tag programmatically
      expect(filterTag).not.toBe(`${baseTag}_action`);
      
      // Note: 'action_movies' contains 'movie' as a substring, but it's a
      // hard-coded string literal, not a concatenation
      expect(filterTag).toBe('action_movies'); // Exact literal match
    });
  });

  describe('Frontend-Backend Tag Consistency', () => {
    it('should match Rust backend tag definitions', () => {
      // These tags must match exactly with src-tauri/src/models.rs tags module
      
      // Base tags (from Rust: pub const SERIES: &str = "series"; etc.)
      expect(HARD_CODED_TAGS.SERIES).toBe('series');
      expect(HARD_CODED_TAGS.MOVIE).toBe('movie');
      expect(HARD_CODED_TAGS.SITCOM).toBe('sitcom');
      expect(HARD_CODED_TAGS.KIDS).toBe('kids');
      expect(HARD_CODED_TAGS.HERO_TRAILER).toBe('hero_trailer');
      
      // Filter tags (from Rust: pub const COMEDY_MOVIES: &str = "comedy_movies"; etc.)
      expect(HARD_CODED_TAGS.COMEDY_MOVIES).toBe('comedy_movies');
      expect(HARD_CODED_TAGS.ACTION_MOVIES).toBe('action_movies');
      expect(HARD_CODED_TAGS.ROMANCE_MOVIES).toBe('romance_movies');
      expect(HARD_CODED_TAGS.COMEDY_SERIES).toBe('comedy_series');
      expect(HARD_CODED_TAGS.ACTION_SERIES).toBe('action_series');
      expect(HARD_CODED_TAGS.ROMANCE_SERIES).toBe('romance_series');
      expect(HARD_CODED_TAGS.COMEDY_KIDS).toBe('comedy_kids');
      expect(HARD_CODED_TAGS.ACTION_KIDS).toBe('action_kids');
    });

    it('should have matching tag counts between frontend and backend', () => {
      // Verify we have exactly 5 base tags and 8 filter tags
      // This matches the Rust backend definition
      expect(BASE_TAGS.length).toBe(5);
      expect(FILTER_TAGS.length).toBe(8);
      
      // Total of 13 tags
      const totalTags = BASE_TAGS.length + FILTER_TAGS.length;
      expect(totalTags).toBe(13);
    });
  });

  describe('Tag System Authoritative Nature', () => {
    it('should use tags from constants, never from external sources', () => {
      // Verify that all category definitions use the hard-coded constants
      const allCategoryTags = [
        CATEGORIES.movies.baseTag,
        ...CATEGORIES.movies.filters.map(f => f.tag),
        CATEGORIES.series.baseTag,
        ...CATEGORIES.series.filters.map(f => f.tag),
        CATEGORIES.sitcoms.baseTag,
        ...CATEGORIES.sitcoms.filters.map(f => f.tag),
        CATEGORIES.kids.baseTag,
        ...CATEGORIES.kids.filters.map(f => f.tag)
      ];

      // Create a set of all valid tags from our constants
      const validTags = new Set<string>([
        ...BASE_TAGS,
        ...FILTER_TAGS
      ]);

      // Every tag in categories should be in our hard-coded lists
      allCategoryTags.forEach(tag => {
        expect(validTags.has(tag)).toBe(true);
      });
    });

    it('should not have any tag inference or suggestion logic', () => {
      // Verify no functions that generate or suggest tags
      // All tag-related functions should only validate or categorize existing tags
      
      // This is a structural test - if these imports exist, they should not
      // contain tag generation logic
      expect(CATEGORIES).toBeDefined();
      expect(HARD_CODED_TAGS).toBeDefined();
      
      // Verify HARD_CODED_TAGS is an object with string values only
      Object.values(HARD_CODED_TAGS).forEach(value => {
        expect(typeof value).toBe('string');
      });
    });

    it('should maintain exact tag strings without normalization', () => {
      // Tags should be used exactly as defined, no case conversion or normalization
      const tags = [
        HARD_CODED_TAGS.SERIES,
        HARD_CODED_TAGS.MOVIE,
        HARD_CODED_TAGS.HERO_TRAILER,
        HARD_CODED_TAGS.COMEDY_MOVIES,
        HARD_CODED_TAGS.ACTION_SERIES
      ];

      tags.forEach(tag => {
        // Tags should be lowercase with underscores (snake_case)
        expect(tag).toBe(tag.toLowerCase());
        expect(tag).not.toContain(' ');
        expect(tag).not.toContain('-');
        
        // If tag contains multiple words, they should be separated by underscores
        if (tag.includes('_')) {
          expect(tag.split('_').length).toBeGreaterThan(1);
        }
      });
    });
  });

  describe('Tag Usage Patterns', () => {
    it('should only use tags from HARD_CODED_TAGS constant', () => {
      // Verify that the application only uses predefined tags
      const validTags = new Set<string>([
        ...BASE_TAGS,
        ...FILTER_TAGS
      ]);

      // All tags in CATEGORIES should be in validTags
      Object.values(CATEGORIES).forEach(category => {
        expect(validTags.has(category.baseTag)).toBe(true);
        
        category.filters.forEach(filter => {
          expect(validTags.has(filter.tag)).toBe(true);
        });
      });
    });

    it('should not allow runtime tag modification', () => {
      // Verify that tag constants are truly constant
      const originalSeriesTag = HARD_CODED_TAGS.SERIES;
      
      // TypeScript prevents modification at compile time
      // Verify the constant value is correct
      expect(HARD_CODED_TAGS.SERIES).toBe(originalSeriesTag);
      expect(HARD_CODED_TAGS.SERIES).toBe('series');
      
      // The HARD_CODED_TAGS object is defined with 'as const'
      // which makes it readonly in TypeScript
    });
  });

  describe('Documentation Compliance', () => {
    it('should match tags documented in design.md', () => {
      // These tags are documented in .kiro/specs/kiyya-desktop-streaming/design.md
      // under "Hard-Coded Tags (Authoritative)" section
      
      // Base Tags from design.md
      const documentedBaseTags = ['series', 'movie', 'sitcom', 'kids', 'hero_trailer'];
      expect([...BASE_TAGS]).toEqual(documentedBaseTags);
      
      // Category Filter Tags from design.md
      const documentedFilterTags = [
        'comedy_movies', 'action_movies', 'romance_movies',
        'comedy_series', 'action_series', 'romance_series',
        'comedy_kids', 'action_kids'
      ];
      expect([...FILTER_TAGS]).toEqual(documentedFilterTags);
    });

    it('should have hero_trailer tag for hero content', () => {
      // Verify hero_trailer tag exists and is used correctly
      expect(HARD_CODED_TAGS.HERO_TRAILER).toBe('hero_trailer');
      expect(BASE_TAGS).toContain('hero_trailer');
      
      // hero_trailer should be a base tag, not a filter tag
      expect(FILTER_TAGS).not.toContain('hero_trailer');
    });
  });
});

File: src-tauri/src/tag_immutability_test.rs
/**
 * Tag System Immutability Verification Tests (Rust Backend)
 * 
 * This test module verifies that the tag system in the Rust backend
 * is immutable and matches the frontend definitions exactly.
 */

#[cfg(test)]
mod tests {
    use crate::models::tags;

    #[test]
    fn test_base_tags_are_constant_strings() {
        // Verify all base tags are defined as constant string slices
        assert_eq!(tags::SERIES, "series");
        assert_eq!(tags::MOVIE, "movie");
        assert_eq!(tags::SITCOM, "sitcom");
        assert_eq!(tags::KIDS, "kids");
        assert_eq!(tags::HERO_TRAILER, "hero_trailer");
    }

    #[test]
    fn test_filter_tags_are_constant_strings() {
        // Verify all filter tags are defined as constant string slices
        // Movie filters
        assert_eq!(tags::COMEDY_MOVIES, "comedy_movies");
        assert_eq!(tags::ACTION_MOVIES, "action_movies");
        assert_eq!(tags::ROMANCE_MOVIES, "romance_movies");
        
        // Series filters
        assert_eq!(tags::COMEDY_SERIES, "comedy_series");
        assert_eq!(tags::ACTION_SERIES, "action_series");
        assert_eq!(tags::ROMANCE_SERIES, "romance_series");
        
        // Kids filters
        assert_eq!(tags::COMEDY_KIDS, "comedy_kids");
        assert_eq!(tags::ACTION_KIDS, "action_kids");
    }

    #[test]
    fn test_base_tags_array_is_immutable() {
        // Verify BASE_TAGS array contains exactly 5 tags
        assert_eq!(tags::BASE_TAGS.len(), 5);
        
        // Verify exact contents
        assert!(tags::BASE_TAGS.contains(&"series"));
        assert!(tags::BASE_TAGS.contains(&"movie"));
        assert!(tags::BASE_TAGS.contains(&"sitcom"));
        assert!(tags::BASE_TAGS.contains(&"kids"));
        assert!(tags::BASE_TAGS.contains(&"hero_trailer"));
    }

    #[test]
    fn test_filter_tags_array_is_immutable() {
        // Verify FILTER_TAGS array contains exactly 8 tags
        assert_eq!(tags::FILTER_TAGS.len(), 8);
        
        // Verify exact contents
        assert!(tags::FILTER_TAGS.contains(&"comedy_movies"));
        assert!(tags::FILTER_TAGS.contains(&"action_movies"));
        assert!(tags::FILTER_TAGS.contains(&"romance_movies"));
        assert!(tags::FILTER_TAGS.contains(&"comedy_series"));
        assert!(tags::FILTER_TAGS.contains(&"action_series"));
        assert!(tags::FILTER_TAGS.contains(&"romance_series"));
        assert!(tags::FILTER_TAGS.contains(&"comedy_kids"));
        assert!(tags::FILTER_TAGS.contains(&"action_kids"));
    }

    #[test]
    fn test_no_dynamic_tag_generation() {
        // Verify that tags are static strings, not computed
        // All tags should be compile-time constants
        
        // Base tags should not be derived from other strings
        let series_tag = tags::SERIES;
        assert_eq!(series_tag, "series");
        assert_eq!(series_tag.len(), 6); // Fixed length, not computed
        
        // Filter tags should not be concatenations
        let action_movies = tags::ACTION_MOVIES;
        assert_eq!(action_movies, "action_movies");
        
        // Verify it's not a runtime concatenation (it's a compile-time constant)
        // The tag is defined as: pub const ACTION_MOVIES: &str = "action_movies";
        // This is a string literal, not format!("{}_{}", "action", "movies")
        let expected_if_concatenated = format!("{}_{}", "action", "movies");
        assert_eq!(action_movies, expected_if_concatenated); // They match, but action_movies is NOT a concatenation
        
        // The key point: action_movies is a &'static str compile-time constant
        // not a String created at runtime
    }

    #[test]
    fn test_tag_validation_functions() {
        // Test is_base_tag function
        assert!(tags::is_base_tag("series"));
        assert!(tags::is_base_tag("movie"));
        assert!(tags::is_base_tag("hero_trailer"));
        assert!(!tags::is_base_tag("invalid_tag"));
        assert!(!tags::is_base_tag("comedy_movies")); // Filter tag, not base
        
        // Test is_filter_tag function
        assert!(tags::is_filter_tag("comedy_movies"));
        assert!(tags::is_filter_tag("action_series"));
        assert!(!tags::is_filter_tag("invalid_tag"));
        assert!(!tags::is_filter_tag("movie")); // Base tag, not filter
    }

    #[test]
    fn test_base_tag_for_filter_mapping() {
        // Test base_tag_for_filter function
        assert_eq!(tags::base_tag_for_filter("comedy_movies"), Some("movie"));
        assert_eq!(tags::base_tag_for_filter("action_movies"), Some("movie"));
        assert_eq!(tags::base_tag_for_filter("romance_movies"), Some("movie"));
        
        assert_eq!(tags::base_tag_for_filter("comedy_series"), Some("series"));
        assert_eq!(tags::base_tag_for_filter("action_series"), Some("series"));
        assert_eq!(tags::base_tag_for_filter("romance_series"), Some("series"));
        
        assert_eq!(tags::base_tag_for_filter("comedy_kids"), Some("kids"));
        assert_eq!(tags::base_tag_for_filter("action_kids"), Some("kids"));
        
        // Invalid filter should return None
        assert_eq!(tags::base_tag_for_filter("invalid_filter"), None);
    }

    #[test]
    fn test_tag_count_consistency() {
        // Verify total tag count matches documentation
        // 5 base tags + 8 filter tags = 13 total
        let total_tags = tags::BASE_TAGS.len() + tags::FILTER_TAGS.len();
        assert_eq!(total_tags, 13);
    }

    #[test]
    fn test_tags_are_lowercase_snake_case() {
        // Verify all tags follow lowercase snake_case convention
        let all_tags = [
            tags::SERIES, tags::MOVIE, tags::SITCOM, tags::KIDS, tags::HERO_TRAILER,
            tags::COMEDY_MOVIES, tags::ACTION_MOVIES, tags::ROMANCE_MOVIES,
            tags::COMEDY_SERIES, tags::ACTION_SERIES, tags::ROMANCE_SERIES,
            tags::COMEDY_KIDS, tags::ACTION_KIDS,
        ];
        
        for tag in &all_tags {
            // Should be lowercase
            assert_eq!(*tag, tag.to_lowercase());
            
            // Should not contain spaces or hyphens
            assert!(!tag.contains(' '));
            assert!(!tag.contains('-'));
            
            // Multi-word tags should use underscores
            if tag.contains('_') {
                assert!(tag.split('_').count() > 1);
            }
        }
    }

    #[test]
    fn test_hero_trailer_is_base_tag() {
        // Verify hero_trailer is a base tag, not a filter tag
        assert!(tags::is_base_tag("hero_trailer"));
        assert!(!tags::is_filter_tag("hero_trailer"));
        assert!(tags::BASE_TAGS.contains(&"hero_trailer"));
        assert!(!tags::FILTER_TAGS.contains(&"hero_trailer"));
    }

    #[test]
    fn test_no_tag_inference_logic() {
        // Verify that the tags module only contains validation functions,
        // not tag generation or inference functions
        
        // The only functions should be:
        // - is_base_tag (validation)
        // - is_filter_tag (validation)
        // - base_tag_for_filter (mapping)
        
        // No functions like:
        // - generate_tag
        // - infer_tag
        // - create_tag
        // - suggest_tag
        
        // This is verified by the fact that all tag constants are &'static str
        // and cannot be modified at runtime
        
        let tag: &'static str = tags::MOVIE;
        assert_eq!(tag, "movie");
    }

    #[test]
    fn test_frontend_backend_tag_consistency() {
        // These tags must match exactly with frontend definitions in:
        // - src/config/categories.ts (HARD_CODED_TAGS)
        // - src/types/index.ts (BASE_TAGS, FILTER_TAGS)
        
        // Base tags
        assert_eq!(tags::SERIES, "series");
        assert_eq!(tags::MOVIE, "movie");
        assert_eq!(tags::SITCOM, "sitcom");
        assert_eq!(tags::KIDS, "kids");
        assert_eq!(tags::HERO_TRAILER, "hero_trailer");
        
        // Filter tags
        assert_eq!(tags::COMEDY_MOVIES, "comedy_movies");
        assert_eq!(tags::ACTION_MOVIES, "action_movies");
        assert_eq!(tags::ROMANCE_MOVIES, "romance_movies");
        assert_eq!(tags::COMEDY_SERIES, "comedy_series");
        assert_eq!(tags::ACTION_SERIES, "action_series");
        assert_eq!(tags::ROMANCE_SERIES, "romance_series");
        assert_eq!(tags::COMEDY_KIDS, "comedy_kids");
        assert_eq!(tags::ACTION_KIDS, "action_kids");
    }
}
